MNT refactoring backend APIs (#1667)
joblib/joblib: doc/custom_parallel_backend.rst

.. code-block:: python

    # Register the backend so it can be used with parallel_config
    register_parallel_backend('custom', MyCustomBackend)

This backend can then be used within the ``parallel_config`` context manager,
as:

.. code-block:: python

    from joblib import Parallel, delayed, parallel_config

    with parallel_config("custom"):
        res = Parallel(2)(delayed(id)(i) for i in range(10))

Minimal backend factory specification
======================================

The backend factory can be any callable that returns an instance of
``ParallelBackendBase``. The complete specification of the backend factory is
referenced in the `default backends source code`_.
Here, we provide minimal details to implement a custom backend factory and
the next section describes various possible customizations.

The key methods for the backend factory are:

- ``configure``: this method is called when creating a particular instance of
  the ``Parallel`` object. It can allocate the resources needed for the
  parallel computation and return the effective number of jobs that can be run
  in parallel.

> - ``effective_n_jobs``: This method is called to determine the number of jobs
>   that can be run in parallel. It can be used to limit the number of jobs that
>   can be run in parallel.

Reviewer: ```suggestion
- ``effective_n_jobs``: This method is called to determine the number of jobs
  that can be run in parallel. It can be used to limit the number of jobs depending
  on the specified ``n_jobs`` and the system it runs on.
```

- ``terminate``: this method is called when the parallel computation is
  finished. It can be used to clean up the resources allocated by the backend.
- ``submit``: this method is called to launch the computation for a given task
  in the ``Parallel`` call. The task is a single callable that takes no
  argument and the method should return a future-like object that allows
  tracking the task's progress. The ``callback`` argument passed to this method

>   is a callable that should be called when the task is completed, typically through the ``add_done_callback`` method of the future-like object.

Reviewer: ```suggestion
  is a callable that should be called when the task is completed, typically through
  the ``add_done_callback`` method of the future-like object.
```

- ``retrieve_result_callback``: This method is called within the callback
  function passed to ``submit``. It is called with the same arguments provided

>   by the backend to the callback functionality. It should retrieve and return the result of the function executed in parallel.

Reviewer: ```suggestion
  by the backend to the callback functionality -- typically the future-like object.
  It should retrieve and return the result of the function executed in parallel.
```

    register_parallel_backend('custom', MyCustomBackend)

The connection parameters can then be passed to the
:func:`~joblib.parallel_config` context manager:

.. code-block:: python

    with parallel_config(backend='custom', endpoint='http://compute',
                         api_key='42'):
        Parallel()(delayed(some_function)(i) for i in range(10))

Using the context manager can be helpful when using a third-party library that
uses :class:`joblib.Parallel` internally while not exposing the ``backend``
argument in its own API.

Cancelling tasks
~~~~~~~~~~~~~~~~

If the backend allow to cancel tasks, the method ``abort_everything`` can be
implemented to abort all the tasks that are currently running as soon as one of
the tasks raises an exception. This can be useful to avoid wasting
computational resources when the call will fail.

This method have an extra parameters ``ensure_ready`` that informs the backend
whether the error was part of a single call to ``Parallel`` or in a context
manager block. In the case of a single call (``ensure_ready=False``), there is
no need to re-spawn workers for future calls, while in the case of a context

> (``ensure_ready=True``),

Reviewer: ```suggestion
(``ensure_ready=True``), one could call ``configure`` to re-allocate computational
resources.
```

threads on the host.

.. code-block:: python

    def get_nested_backend(self):
        """Backend instance to be used by nested Parallel calls.

        By default a thread-based backend is used for the first level of
        nesting. Beyond, switch to sequential backend to avoid spawning too
        many threads on the host.
        """
        nesting_level = getattr(self, "nesting_level", 0) + 1
        return LokyBackend(nesting_level=nesting_level), None

Another nested parallelism that needs to be controlled is the numbers of thread
in third-party C-level threadpools, *e.g.* OpenMP, MKL, or BLAS. In ``joblib``,
this is controlled with the ``inner_max_num_threads`` argument that can be
provided to the backend in the ``parallel_config`` context manager. To support
this argument, the backend should set the ``supports_inner_max_num_threads``
class attribute to ``True`` and accept the argument in the constructor to set
this up in the workers. A helper to set this in the workers is to use
environment variables provided by ``self._prepare_worker_env(n_jobs)``.

Third-party backend registration
================================

> A problem exists that external packages that register new parallel backends
> must now be imported explicitly for their backends to be identified by joblib::

Reviewer: ```suggestion
To be used seamlessly within :func:`~joblib.parallel_config`, external packages need
to register their parallel backends so they are identified by joblib. This requires
importing extra packages as in::
```
